[DCP] Enable filesystem/fsspec auto detection

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119287
* __->__ #118888

This API enables the ability to automatically detect whether to use filesystem or fsspec based on the checkpoint_id.

Differential Revision: [D53318043](https://our.internmc.facebook.com/intern/diff/D53318043/)

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @LucasLLC

fegin committed:
[DCP] Enable filesystem/fsspec auto detection

This API enables the ability to automatically detect whether to use filesystem or fsspec based on the checkpoint_id.

Differential Revision: [D53318043](https://our.internmc.facebook.com/intern/diff/D53318043/)

[ghstack-poisoned]

Skylion007 commented on torch/distributed/checkpoint/_fsspec_filesystem.py:
```diff
@@ -50,6 +51,18 @@ def rename(
         self.fs.makedirs(path, exist_ok=True)
 
+    @staticmethod
+    def check(path: Union[str, os.PathLike, None] = None) -> bool:
```
fsspec handles local files fine though? What's the motivation here? You can give unprefixed files to fsspect and it will just return a LocalFileSystem, fsspec object?

fegin replied:
The original motivation is to keep the filesystem reader and writer as the default one as we don't know how fsspec can affect the performance. Also, there was also concern that taking fsspec as the hard dependency. But it is added to `requirements.py`. So it may not be the case anymore.

fegin committed:
Update on "[DCP] Enable filesystem/fsspec auto detection"


This API enables the ability to automatically detect whether to use filesystem or fsspec based on the checkpoint_id.

Differential Revision: [D53318043](https://our.internmc.facebook.com/intern/diff/D53318043/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 LucasLLC

[ghstack-poisoned]

Skylion007 commented on torch/distributed/checkpoint/storage_utils.py:
```diff
@@ -20,6 +20,7 @@ def _storage_setup(
             "storage_reader/storage_writer is None."
         )
 
+    targets: List[type[Union[StorageReader, StorageWriter]]] = []
```
might want to stick with typing.Type for now for checkers on older version of Python

fegin committed:
Update on "[DCP] Enable filesystem/fsspec auto detection"


This API enables the ability to automatically detect whether to use filesystem or fsspec based on the checkpoint_id.

Differential Revision: [D53318043](https://our.internmc.facebook.com/intern/diff/D53318043/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 LucasLLC

[ghstack-poisoned]

fegin committed:
Update on "[DCP] Enable filesystem/fsspec auto detection"


This API enables the ability to automatically detect whether to use filesystem or fsspec based on the checkpoint_id.

Differential Revision: [D53318043](https://our.internmc.facebook.com/intern/diff/D53318043/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 LucasLLC

[ghstack-poisoned]

LucasLLC commented on test/distributed/checkpoint/test_save_load_api.py:
```diff
@@ -0,0 +1,65 @@
+    def test_auto_detect(self):
+        model = FSDP(MyTestModule().cuda())
+        device_mesh = init_device_mesh(self.device_type, (self.world_size,))
+        model = FSDP(model, device_mesh=device_mesh)
```
Just curious why the model is wrapped w/ FSDP twice?

fegin replied:
It's a mistake. I'll fix this in a separate PR.

wz337 reviewed (approved):
Stamped from internal.

LucasLLC commented on torch/distributed/checkpoint/_storage_utils.py:
```diff
@@ -0,0 +1,48 @@
+import os
+from typing import List, Type, Union
+
+from .filesystem import FileSystemReader, FileSystemWriter
```
nit: I tend to avoid relative imports since it's not PEP8 friendly.

LucasLLC commented on test/distributed/checkpoint/test_checkpoint.py:
```diff
@@ -198,6 +198,10 @@ def write_data(
     def finish(self, metadata: Metadata, results: List[List[WriteResult]]) -> None:
         self._fail_rank("fail_finish")
 
+    @staticmethod
```
nit: I tend to favor @classmethod in instances such as these since it has better OO support.

LucasLLC commented on test/distributed/checkpoint/test_checkpoint.py:
```diff
@@ -198,6 +198,10 @@ def write_data(
         self._fail_rank("fail_finish")
 
+    @staticmethod
+    def check(checkpoint_id: Union[str, os.PathLike]) -> bool:
```
nit: `check` could be considered a little vague, what about something like `is_valid_checkpoint_id` or `is_valid`?

LucasLLC reviewed (approved):
A couple nit's for your consideration but LGTM thanks :)

fegin committed:
Update on "[DCP] Enable filesystem/fsspec auto detection"


This API enables the ability to automatically detect whether to use filesystem or fsspec based on the checkpoint_id.

Differential Revision: [D53318043](https://our.internmc.facebook.com/intern/diff/D53318043/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 LucasLLC

[ghstack-poisoned]

fegin committed:
Update on "[DCP] Enable filesystem/fsspec auto detection"


This API enables the ability to automatically detect whether to use filesystem or fsspec based on the checkpoint_id.

Differential Revision: [D53318043](https://our.internmc.facebook.com/intern/diff/D53318043/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 LucasLLC

[ghstack-poisoned]

ppwwyyxx commented on torch/distributed/checkpoint/filesystem.py:
```diff
@@ -358,6 +363,20 @@ def rename(
+        if "://" in str(checkpoint_id):
+            return False
+
+        for p in Path(checkpoint_id).parents:
+            if p.exists() and os.access(str(p), os.W_OK):
+                return True
```
Why is it necessary to require the path to be writable?
If the path is read-only, users should be allowed to use DCP to read checkpoints.
